[ews] Add support for accepting build data from generic CI builds in ews django app

aj062 opened the pull request:
#### 70b4f85956c975524b3cd17302033b2621b6c444
<pre>
[ews] Add support for accepting build data from generic CI builds in ews django app
<a href="https://bugs.webkit.org/show_bug.cgi?id=296197">https://bugs.webkit.org/show_bug.cgi?id=296197</a>
<a href="https://rdar.apple.com/156161841">rdar://156161841</a>

Reviewed by Brianna Fan.

Add support to receive build results from generic CI builds, so that it can receive and
process the data and save it in the database appropriately.

* Tools/CISupport/ews-app/ews/views/results.py:
(Results.post):
(Results.ci_build_event):
(Results.build_event):
(Results.step_event):
(Results.is_valid_result):

Canonical link: <a href="https://commits.webkit.org/297677@main">https://commits.webkit.org/297677@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/707b46f26ba1513a4bb454fd7fb36e6c55423dcb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112546 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/32278 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/22756 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/118744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/63000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114508 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/32930 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/40841 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/118744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/63000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115493 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/32930 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/22756 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/118744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/32930 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/22756 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62504 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/32930 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/22756 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121966 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/39620 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/40841 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/40002 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/22756 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/94263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/22756 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18122 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/39508 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/39145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/42480 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/40886 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/CISupport/ews-app/ews/views/results.py:
```diff
@@ -50,19 +51,39 @@ def post(self, request):
+        pr_number = data['pr_number']
+        pr_project = data.get('pr_project', 'WebKit/WebKit')
+        result = data['result']
+        url = data.get('url') or 'https://apple.com'
```
Do we want to keep the testing URL here? Or should we make it none?

aj062 replied:
Having url as '' or None was causing some issues (e.g.: having url as '' breaks Github markdown syntax making status-bubbles text garbage). 
We don't allow null in the model, django doesn't recommend allowing null as a value for TextField (https://docs.djangoproject.com/en/5.2/ref/models/fields/#null).

So this is a simple way to ensure we don't have those values.

briannafan commented on Tools/CISupport/ews-app/ews/views/results.py:
```diff
@@ -82,26 +103,24 @@ def build_event(self, data):
                                            'number', 'result', 'state_string', 'started_at', 'complete_at'],
-                         u'ews-step': ['hostname', 'step_id', 'build_id', 'result', 'state_string', 'started_at', 'complete_at']}
+                         'ews-step': ['hostname', 'step_id', 'build_id', 'result', 'state_string', 'started_at', 'complete_at'],
+                         'ci-build': ['build_id', 'builder_display_name', 'commit_hash', 'pr_number', 'result', 'url']}
```
Oh, I see this is what ensures we have a pr number in https://github.com/WebKit/WebKit/pull/48253/files#r2216751022, right?

aj062 replied:
Right!

briannafan commented on Tools/CISupport/ews-app/ews/views/results.py:
```diff
@@ -50,19 +51,39 @@ def post(self, request):
-        if data.get('type') == u'ews-build':
+        data.pop('EWS_API_KEY', None)
+
+        if not self.is_valid_result(data):
```
Good move! It makes a lot more sense for this check to be here.

briannafan commented on Tools/CISupport/ews-app/ews/views/results.py:
```diff
@@ -50,19 +51,39 @@ def post(self, request):
+        result = data['result']
+        url = data.get('url') or 'https://apple.com'
+
+        rc = CIBuild.save_build(commit_hash=commit_hash, build_id=data['build_id'], builder_display_name=data['builder_display_name'],
```
nit: Can we retrieve the build_id and builder_display_name in the same way as the other data?

aj062 committed:
[ews] Add support for accepting build data from generic CI builds in ews django app
https://bugs.webkit.org/show_bug.cgi?id=296197
rdar://156161841

Reviewed by Brianna Fan.

Add support to receive build results from generic CI builds, so that it can receive and
process the data and save it in the database appropriately.

* Tools/CISupport/ews-app/ews/views/results.py:
(Results.post):
(Results.ci_build_event):
(Results.build_event):
(Results.step_event):
(Results.is_valid_result):

Canonical link: https://commits.webkit.org/297677@main
